File: python/20_modules.py
# ============================================================
# Topic:  Modules & Imports  (FOCUS: import, from-import, alias, __name__)
# Summary: Organize code into modules; stdlib import patterns.
# Run:     python python/20_modules.py
#
# KEY DIFFERENCES:
#   Python:  import / from … import / as alias + sys.path
#   TS/JS:   import / export (ES modules) | require (CommonJS)
#   Rust:    mod + use + crate system
#   Java:    package + import
#   C++:     #include + namespaces
# ============================================================

import json
import math
import sys
import types
from math import sqrt
import math as m


# ----- Package layout (multi-file) -----
# mypkg/
#   __init__.py      def greet(): return "Hi"
#   utils.py         helper functions
# from mypkg import greet


# ----- 1. Import module — use math and json -----
def demo_import_module():
    data = json.dumps({"pi": round(math.pi, 2)})
    parsed = json.loads(data)
    print(f"pi: {parsed['pi']:.2f}", flush=True)  # Output: pi: 3.14


# ----- 2. from-import — pull a name from a module -----
_demo_pkg = types.ModuleType("demo_pkg")
_demo_pkg.greet = lambda: "Hi"
sys.modules["demo_pkg"] = _demo_pkg


def demo_from_import():
    from demo_pkg import greet

    print(f"greet: {greet()}", flush=True)  # Output: greet: Hi


# ----- 3. Alias — import math as m; from math import sqrt -----
def demo_alias():
    print(f"sqrt: {sqrt(25)}", flush=True)  # Output: sqrt: 5.0
    print(f"alias: {int(m.pow(5, 2))}", flush=True)  # Output: alias: 25


# ----- 4. __name__ — script vs imported module -----
def demo_name():
    if __name__ == "__main__":
        print("main module", flush=True)  # Output: main module


# ----- 5. sys.path note -----
# sys.path lists directories Python searches for modules (cwd, stdlib, site-packages).
# Add a path with sys.path.insert(0, "/my/project") or set PYTHONPATH.


def main():
    # ----- 1. Import module -----
    demo_import_module()

    # ----- 2. from-import -----
    demo_from_import()

    # ----- 3. Alias -----
    demo_alias()

    # ----- 4. __name__ -----
    demo_name()


if __name__ == "__main__":
    main()

# Cross-language note:
#   Python modules are .py files; packages are folders with __init__.py.
#   __name__ == "__main__" guards script-only code (like Java public static void main).
#   JS/TS use export/import; Rust uses mod/use; Java uses package/import; C++ uses #include.

File: cpp/20_modules.cpp
// ============================================================
// Topic:  Modules & Namespaces  (FOCUS: #include, namespace, using, alias)
// Summary: Header includes + namespace util — C++20 modules optional; headers universal.
// Run:     g++ -std=c++17 cpp/20_modules.cpp -o cpp/20_modules && cpp/20_modules
//
// KEY DIFFERENCES:
//   C++:     #include headers, namespace {}, using, namespace alias (C++17 has no modules by default)
//   Python:  import module, from x import y, as alias
//   Java:    import pkg.Class, packages as namespaces
//   C#:      using Namespace; using Alias = Full.Type;
//   Rust:    mod / use / pub mod, crate hierarchy
//   TS/JS:   import / export (ES modules)
// ============================================================

#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

// ----- util namespace — groups related free functions and constants -----
namespace util {

constexpr double kPi = 3.141592653589793;

std::string greet(const std::string& name) { return "Hi, " + name; }

double square(double value) { return value * value; }

}  // namespace util

// ----- using declaration — bring single name into scope -----
using util::greet;

// ----- namespace alias — shorthand for qualified names -----
namespace u = util;

int main() {
    // ----- 1. util::kPi via namespace -----
    std::cout << "pi: " << std::fixed << std::setprecision(2) << util::kPi << std::endl;
    // Output: pi: 3.14

    // ----- 2. greet via using declaration (no util:: prefix) -----
    std::cout << "greet: " << greet("World").substr(0, 2) << std::endl;
    // Output: greet: Hi

    // ----- 3. <cmath> — std::sqrt from included header -----
    std::cout << "sqrt: " << std::fixed << std::setprecision(1) << std::sqrt(25.0) << std::endl;
    // Output: sqrt: 5.0

    // ----- 4. namespace alias u::square -----
    std::cout << "alias: " << static_cast<int>(u::square(5.0)) << std::endl;
    // Output: alias: 25

    std::cout << "main module" << std::endl;
    // Output: main module

    return 0;
}